src/push-notifications.js

```diff
@@ -1,17 +1,140 @@
+  return new Client(config);
+}
+
+export class Client {
+  constructor(config) {
+    if (!config) {
+      throw new Error('Config object required');
+    }
+    const { instanceId, endpointOverride = null } = config;
+
+    if (instanceId === undefined) {
+      throw new Error('Instance ID is required');
+    }
+    if (typeof instanceId !== 'string') {
+      throw new Error('Instance ID must be a string');
+    }
+    if (instanceId.length === 0) {
+      throw new Error('Instance ID cannot be empty');
+    }
+
+    this.instanceId = instanceId;
+    this._endpoint = endpointOverride; // Internal only
+
+    this._initDb('beams');
+  }
+
+  get _baseURL() {
+    if (this._endpoint !== null) {
+      return this._endpoint;
+    }
+    return `https://${this.instanceId}.pushnotifications.pusher.com`;
+  }
+
+  async start() {
+    const { vapidPublicKey: publicKey } = await this._getPublicKey();
+
+    // register with pushManager, get endpoint etc
+    const token = await this._getPushToken(publicKey);
+
+    // get device id from errol
+    const response = await this._registerDevice(token);
+    // // put response.id in indexedDB
+    this.deviceId = response;
+
+    this._save(this.instanceId, token, this.deviceId);
+  }
+
+  async _getPublicKey() {
+    const path = `${this._baseURL}/device_api/v1/instances/${encodeURIComponent(
+      this.instanceId
+    )}/web-vapid-public-key`;
+    return doRequest('GET', path);
   }
-  function start() {
-    console.log('starting');
+
+  async _getPushToken(publicKey) {
+    let sub;
+    try {
+      window.navigator.serviceWorker.register('sw.js');
+      const reg = await window.navigator.serviceWorker.ready;
+      sub = await reg.pushManager.subscribe({
+        userVisibleOnly: true,
+        applicationServerKey: urlBase64ToUInt8Array(publicKey),
+      });
+    } catch (e) {
+      return Promise.reject(e);
+    }
+
+    return btoa(JSON.stringify(sub));
+  }
+
+  async _registerDevice(token) {
+    const path = `${this._baseURL}/device_api/v1/instances/${encodeURIComponent(
+      this.instanceId
+    )}/devices/web`;
+
+    return doRequest('POST', path, { token });
   }
 
-  function setUserId() {
-    console.log('hey');
+  _initDb(dbName) {
+    const request = indexedDB.open(dbName);
+
+    request.onerror = event => {
+      console.error(`Database error: ${event.target.errorCode}`);
+    };
+
+    request.onsuccess = event => {
+      const db = event.target.result;
+    };
+
+    request.onupgradeneeded = event => {
+      const db = event.target.result;
+      var objectStore = db.createObjectStore('beams', {
+        keyPath: 'instance_id',
+      });
+      objectStore.createIndex('instance_id', 'instance_id', { unique: true });
+      objectStore.createIndex('token', 'token', { unique: true });
+      objectStore.createIndex('device_id', 'device_id', { unique: true });
+    };
   }
 
-  return {
-    start,
-    setUserId
-  };
+  _save(instanceId, token, deviceId) {
```

Comment:
On second thought, we should handle for the case where the browser doesn't support indexedDB (might be an old version of Chrome)

Reply:
Done in b6b6d281d820e72df861b4cdc31b5a8fbb1d4a2b.